File: Data Structure/Tree/Union Find/union-find.js
class UnionFind {

    constructor(count) {
        this.count  = count;
        this.parent = new Array(count);
        this.rank   = new Array(count);
        
        for (let i = 0; i < count; ++i) {
            this.parent[i] = i;
            this.rank[i]   = 1;
        }
    }

    find(p) {
        while (p != this.parent[p]) {
            this.parent[p] = this.parent[this.parent[p]];
            p = this.parent[p];
        }
        return p;
    }

    union(p, q) {
        let rootP = this.find(p),
            rootQ = this.find(q);
        
        if (rootP == rootQ) return;

        let rankP = this.rank[p],
            rankQ = this.rank[q];
        
        if (rankP < rankQ) this.parent[rootP] = rootQ;
        else if (rankP > rankQ) this.parent[rootQ] = rootP;
        else {
            this.parent[rootP] = rootQ;
            this.rank[rootP] += 1;
        }
    }

    isConnected(p, q) {
        return this.find(p) == this.find(q);
    }
} 

File: Data Structure/Tree/Union Find/union_find.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank   = [1] * n

    def find(self, p):
        if p != self.parent[p]:
            self.parent[p] = self.find(self.parent[p])
        return self.parent[p]

    def union(self, p, q):
        root_p, root_q = self.find(p), self.find(q)

        if root_p == root_q:
            return False
        
        rank_p, rank_q = self.rank[p], self.rank[q]
        if rank_p < rank_q:
            self.parent[root_p] = root_q
        elif rank_p > rank_q:
            self.parent[root_q] = root_p
        else:
            self.parent[root_p] = root_q
            self.rank[root_p] += 1
        
        return True
        
    def is_connected(self, p, q):
        return self.find(p) == self.find(q) 
    